File: gui/src/Network.cpp
/*
** EPITECH PROJECT, 2023
** net
** File description:
** et
*/

#include "Network.hpp"
#include <sys/select.h>

Network::Network(char *ip, int port) : _ip(ip), _port(port)
{
    struct sockaddr_in server;
    _sock = socket(AF_INET, SOCK_STREAM, 0);
    if (_sock < 0)
    {
        std::cout << "Socket failed" << std::endl;
        exit(84);
    }
    server.sin_addr.s_addr = inet_addr(ip);
    server.sin_family = AF_INET;
    server.sin_port = htons(port);
    if (connect(_sock, (struct sockaddr *)&server, sizeof(server)) < 0)
    {
        std::cout << "Connection failed" << std::endl;
        exit(84);
    }
}

Network::~Network()
{
    close(_sock);
}

void Network::send_data(std::string data)
{
    send(_sock, data.c_str(), data.length(), 0);
}

// faire select

std::string Network::receive_data()
{
    int size_str = 0, len = 0;
    char buff[4096];
    memset(buff, 0, 4096);
    std::string data = "";
    while (size_str == 0 || buff[size_str - 1] != '\n')
    {
        FD_ZERO(&_readfds);
        FD_SET(_sock, &_readfds);
        select(_sock + 1, &_readfds, NULL, NULL, NULL);
        len = read(_sock, buff + size_str, 1);
        if (len == 0)
        {
            std::cout << "Server disconnected" << std::endl;
            close(_sock);
            throw std::runtime_error("Server disconnected");
        }
        else if (len < 0)
        {
            usleep(1000);
            if (errno == 9)
            {
                std::cout << "Server disconnected" << std::endl;
                close(_sock);
                throw std::runtime_error("Server disconnected");
            }
        }
        else
        {
            size_str += len;
        }
    }
    if (len < 0 || buff[0] == '\0')
        return "";
    data = buff;
    return data;
}

File: ai/src/network.py
##
## EPITECH PROJECT, 2023
## B-YEP-400-MAR-4-1-zappy-kenan.blasius
## File description:
## network
##

import socket
import sys
import select

from logger import Logger

class Network:
    def __init__(self, data):
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.server.connect((data["host"], data["port"]))
        except ConnectionRefusedError:
            Logger.log_err("Error: Connection refused to: " + data["host"] + ":" + str(data["port"]))
            sys.exit(84)

        self.buffer = ""

    def send(self, message):
        self.server.send((message + "\n").encode())

    def recv(self):
        while "\n" not in self.buffer:
            try:
                select.select([self.server], [], [], 0)
            except select.error:
                return exit(84)
            self.buffer += self.server.recv(1024).decode()
        message = self.buffer.split("\n")[0]
        self.buffer = self.buffer[len(message) + 1:]
        return message

    def close(self):
        self.server.close()
